src/menus/EditMenus.cpp

```diff
@@ -73,65 +79,173 @@ bool DoPasteText(AudacityProject &project)
    return false;
 }
 
-// Return true if nothing selected, regardless of paste result.
-// If nothing was selected, create and paste into NEW tracks.
+void ForEachCopiedWaveTrack(const TrackList& src,
+                            const TrackList& dst,
+                            const std::function<void(const WaveTrack& waveTrack)>& f)
+{
+   if(dst.Selected().empty())
+   {
+      for(auto waveTrack : src.Any<const WaveTrack>())
+         f(*waveTrack);
+   }
+   else
+   {
+      const auto srcTrackRange = src.Any<const WaveTrack>();
+      const auto dstTrackRange = dst.Any<const WaveTrack>();
+      auto srcTrack = srcTrackRange.begin();
+      auto dstTrack = dstTrackRange.begin();
+      auto lastCopiedTrack = srcTrack;
+      while(dstTrack != dstTrackRange.end() && srcTrack != srcTrackRange.end())
+      {
+         if(!(*dstTrack)->GetSelected())
+         {
+            ++dstTrack;
+            continue;
+         }
+         
+         auto srcChannelCount = TrackList::Channels(*srcTrack).size();
+         auto dstChannelCount = TrackList::Channels(*dstTrack).size();
+                  
+         while(srcChannelCount > 0 && dstChannelCount > 0)
+         {
+            f(**srcTrack);
+            
+            lastCopiedTrack = srcTrack;
+            ++srcTrack;
+            ++dstTrack;
+            --srcChannelCount;
+            --dstChannelCount;
+         }
+         
+         while(dstChannelCount > 0)
+         {
+            f(**lastCopiedTrack);
+            ++dstTrack;
+            --dstChannelCount;
+         }
+      }
+      while(dstTrack != dstTrackRange.end())
+      {
+         if((*dstTrack)->GetSelected())
+            f(**lastCopiedTrack);
+         ++dstTrack;
+      }
```

Comment:
But what about the other possibility, that destination tracks ran out before source tracks?  Don't you need to check that and visit extra source tracks?  You do that at the top of the function, only when there are zero destination tracks.

Reply:
When copying tracks we consider two cases when pasting:
1) There is no selection
2) Not empty region is selected
In the first case we copy **all tracks** from `src` (used in simplified paste method `DoPasteNothingSelected`). When selection isn't empty `N = min(src.size(), dst.size())` tracks are copied from `src`, plus the **last track** from `src` could be duplicated `M = dst.size() - N` times more, if `M > 0` (corresponds to a paste logic after `!tracks.Selected()` condition in `OnPaste`). In both cases `ForEachCopiedWaveTrack` visits tracks that are to be copied according to behaviour described above.
Does that answers your questions?

Reply:
Yes, I accept this reasoning.  But then the new function should have a comment containing this explanation.  Reading the function in isolation made me question it.